### DISCUSSION TITLE
What is the best way to apply changed theme options?

### DISCUSSION THREAD

**Discussion by moving-bits** [mapsforge/vtm#924] [Q&A] [answered]
I've reaching theming in my preparation for a switch from Mapsforge to VTM. Selecting a theme does work, displaying, selecting and storing theme options as well. Everything seems to be fine, but when I close the theme options menu (by either left arrow in the toolbar or back button), the selected changes are not applied immediately, but only on restarting the map. Is there some best practice on how to apply changed options? Thanks.

**Comment by moving-bits** [comment]
Found a way that seems to work - `map.setTheme` in map activity's `onResume()` -, but I guess there must be a less costly way to achieve the desired results.

**Comment by devemux86** [comment, marked as answer]
Please check the various samples:

- [MapsforgeStyleActivity](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/MapsforgeStyleActivity.java) and [MapsforgeStyleTest](https://github.com/mapsforge/vtm/blob/master/vtm-playground/src/org/oscim/test/MapsforgeStyleTest.java) show how to use the [map layers](https://github.com/mapsforge/vtm/blob/master/docs/Rendertheme.md#stylemenus).
- Desktop vector samples can change the themes via [1-5] keys, see [InputHandler](https://github.com/mapsforge/vtm/blob/master/vtm-gdx/src/org/oscim/gdx/InputHandler.java).

**Reply by moving-bits** [reply]
Thanks for your reply and for providing the links. (I should have added that this is about c:geo, and therefore Android app, in preparation for a possible change from Mapsforge "classic" to Mapsforge VTM.)

Changing the theme in its entirety does work, as well as displaying the theme options menu does. (It's implemented in a way very similar to the examples.) Also storing changed theme options and applying them on a restart of the map does work. 

The only thing that did not work was applying changed theme options on closing the theme options menu. I "fixed" this by applying a `setTheme(...)` in the `onResume` method of the map activity, which gets called on closing the theme options menu (and many other occasions), but that seems to be a bit costly to me. I would have thought, that the connection between `XmlRenderThemeStyleMenu` and the map's layers would update styling automatically on changing a theme option.

I'm not sure how the samples provided will help me when working with theme options - but maybe I'm missing something:
- The `MapsforgeStyleActivity` uses `setTheme` as well on changing a theme option (here: enabling/disabling nature layer).
- `MapsforgeStyleTest` changes style definitions "on the fly" - that's not how far I do want to go, I'm only using the style options already provided by the theme (OpenAndroMap's "Elevate" theme, in this case).
- `InputHandler` example changes the whole theme (using `setTheme` as well), not some theme options.

So maybe using `setTheme` actually _is_ the way to go, but I can't remember (and can't find) having to use this for Mapsforge (non-VTM).

**Comment by devemux86** [comment]
VTM / Mapsforge API is the same on Android + Dekstop.

Mapsforge map layers are demonstrated in [StyleMenuMapViewer](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-samples-android/src/main/java/org/mapsforge/samples/android/StyleMenuMapViewer.java) (combined with [Settings](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-samples-android/src/main/java/org/mapsforge/samples/android/Settings.java)).

However each map library has different engine and API.
